src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1368,52 +1435,378 @@ private sealed class RemainingTestsForCase
+                    out Tests whenTrue,
+                    out Tests whenFalse,
+                    ref bool foundExplicitNullTest)
                 {
-                    case BoundKind.DagTypeTest:
-                        return ((BoundDagTypeTest)x).Type.Equals(((BoundDagTypeTest)y).Type, TypeCompareKind.AllIgnoreOptions);
+                    whenTrue = whenFalse = this;
+                }
+            }
 
-                    case BoundKind.DagValueTest:
-                        return ((BoundDagValueTest)x).Value == ((BoundDagValueTest)y).Value;
+            /// <summary>
+            /// A single test to be performed, described by a <see cref="BoundDagTest"/>.
+            /// Note that the test might be a <see cref="BoundDagEvaluation"/>, in which case it is deemed to have
+            /// succeeded after being evaluated.
+            /// </summary>
+            public class One : Tests
+            {
+                public readonly BoundDagTest Test;
+                public One(BoundDagTest Test) => this.Test = Test;
+                public void Deconstruct(out BoundDagTest Test) => Test = this.Test;
+                public override void Filter(
+                    DecisionDagBuilder builder,
+                    BoundDagTest test,
+                    IValueSet? whenTrueValues,
+                    IValueSet? whenFalseValues,
+                    out Tests whenTrue,
+                    out Tests whenFalse,
+                    ref bool foundExplicitNullTest)
+                {
+                    builder.CheckConsistentDecision(
+                        test: test,
+                        other: Test,
+                        whenTrueValues: whenTrueValues,
+                        whenFalseValues: whenFalseValues,
+                        syntax: test.Syntax,
+                        trueTestPermitsTrueOther: out bool trueDecisionPermitsTrueOther,
+                        falseTestPermitsTrueOther: out bool falseDecisionPermitsTrueOther,
+                        trueTestImpliesTrueOther: out bool trueDecisionImpliesTrueOther,
+                        falseTestImpliesTrueOther: out bool falseDecisionImpliesTrueOther,
+                        foundExplicitNullTest: ref foundExplicitNullTest);
+                    whenTrue = trueDecisionImpliesTrueOther ? Tests.True.Instance : trueDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
+                    whenFalse = falseDecisionImpliesTrueOther ? Tests.True.Instance : falseDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
+                }
+                public override BoundDagTest ComputeSelectedTest() => this.Test;
+                public override Tests RemoveEvaluation(BoundDagEvaluation e) => Test is BoundDagEvaluation e2 && e2 == e ? Tests.True.Instance : (Tests)this;
+                public override string Dump(Func<BoundDagTest, String> dump) => dump(this.Test);
+                public override bool Equals(object? obj)
```

Comment:
>Equals [](start = 37, length = 6)

Where is `Tests.Equals()` used?

Reply:
See response on `Tests.GetHashCode`

---
In reply to: [382857358](https://github.com/dotnet/roslyn/pull/41503#discussion_r382857358) [](ancestors = 382857358)